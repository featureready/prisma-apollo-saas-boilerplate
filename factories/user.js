export default ({ id }) => ({
  id,
  firstName: 'John',
  lastName: 'Kueh',
  email: 'john@beaconmaker.com',
  createdAt: new Date(1988, 2, 2),
  updatedAt: new Date(1988, 2, 3),
  periodStart: new Date(1988, 10, 10),
  periodEnd: new Date(1988, 10, 11),
  team: {
    id: 1,
    name: 'A team name'
  }
});
