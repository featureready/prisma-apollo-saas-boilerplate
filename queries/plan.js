export const PLANS = `
  query {
    Plans {
      id
      amount
      currency
      interval
      title
      description
    }
  }
`;

export default {
  PLANS
};
