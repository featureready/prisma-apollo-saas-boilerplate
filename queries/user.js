import { gql } from 'apollo-server-express';

export const ME = gql`
  query {
    Me {
      id
      firstName
      lastName
      fullName
      email
      createdAt
      updatedAt
      periodStart
      periodEnd
      team {
        id
        name
      }
    }
  }
`;

export const PAYMENT_HISTORY = `
query {
  PaymentHistory {
    amountDue
    amountPaid
    invoicePdf
    status
    date
    periodStart
    periodEnd
  }
}
`;

export const SIGNUP = `
mutation Signup($input: SignupInput!) {
  Signup(input: $input) {
    jwt
    user {
      id
      firstName
      lastName
      email
    }
  }
}
`;

export const LOGIN = `
mutation Login($input: LoginInput!) {
  Login(input: $input) {
    jwt
    user {
      id
      firstName
      lastName
      email
    }
  }
}
`;

export const UPDATE_USER = `
mutation UpdateUser($input: UpdateUserInput!) {
  UpdateUser(input: $input) {
    id
    firstName
    lastName
    email
  }
}
`;

export const FORGOT_PASSWORD = `
mutation ForgotPassword($input: ForgotPasswordInput!) {
  ForgotPassword(input: $input) {
    message
  }
}
`;

export const RESET_PASSWORD = `
mutation ResetPassword($input: ResetPasswordInput!) {
  ResetPassword(input: $input) {
    message
  }
}
`;

export const DELETE_USER = `
mutation DeleteUser {
  DeleteUser {
    id
  }
}
`;

export const ADD_CREDIT_CARD = `
mutation AddCreditCard($input: AddCreditCardInput!) {
  AddCreditCard(input: $input) {
    message
  }
}
`;

export const SUBSCRIBE_PLAN = `
mutation SubscribePlan($input: SubscribePlanInput!) {
  SubscribePlan(input: $input) {
    message
  }
}
`;

export const INVITE_USER = `
mutation InviteUser($input: InviteUserInput!) {
  InviteUser(input: $input) {
    email
    firstName
    lastName
    status
  }
}
`;
