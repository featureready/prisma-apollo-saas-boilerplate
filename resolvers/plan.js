import _ from 'lodash';
import { listAllPlans } from '../services/stripe';
import { toCamelCase } from '../helpers/arrayUtils';

export default {
  Query: {
    async Plans(parent, args, context, info) {
      const plans = await listAllPlans();

      return toCamelCase(plans);
    }
  }
};
