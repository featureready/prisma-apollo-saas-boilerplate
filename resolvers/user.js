import bcrypt from 'bcrypt';
import jsonwebtoken from 'jsonwebtoken';
import uuidv4 from 'uuid/v4';
import { UserInputError, AuthenticationError } from 'apollo-server';
import sendEmail from '../services/sendgrid';
import {
  createCustomer,
  createCard,
  createSubscription,
  listAllInvoices
} from '../services/stripe';
import { toCamelCase } from '../helpers/arrayUtils';

export default {
  Query: {
    async me(parent, args, { user, prisma }, info) {
      return prisma.user({ id: user.id });
    },
    async paymentHistory(parent, args, context, info) {
      const { stripeCustomerId: customerId } = context.user;
      const invoices = await listAllInvoices({
        customerId
      });

      return toCamelCase(invoices);
    }
  },
  Mutation: {
    async signup(parent, { firstName, lastName, email, password }, { prisma }, info) {
      if (await prisma.user({ email })) {
        throw new UserInputError('Email is already taken');
      } else {
        const user = await prisma.createUser({
          firstName,
          lastName,
          email,
          password: hashedPassword(password),
          stripeCustomerId: await createCustomer({ email })
        });

        return {
          user,
          jwt: getJwt({
            id: user.id,
            email: user.email
          })
        };
      }
    },
    async login(parent, { email, password }, { prisma }, info) {
      const user = await prisma.user({ email });

      if (user && bcrypt.compareSync(password, user.password)) {
        return {
          user,
          jwt: getJwt({
            id: user.id,
            email: user.email
          })
        };
      }

      throw new AuthenticationError('Please check your credentials and try again.');
    },
    async forgotPassword(parent, { email }, { prisma }, info) {
      const user = await prisma.user({ email });

      if (user) {
        const { email: userEmail } = user;
        const token = uuidv4();

        await prisma.updateUser({
          where: { email },
          data: { resetPasswordToken: token }
        });

        sendEmail({
          template_id: process.env.FORGOT_PASSWORD_TEMPLATE_ID,
          to: userEmail,
          from: process.env.SUPPORT_EMAIL_ADDRESS,
          dynamic_template_data: {
            email: userEmail,
            resetPasswordLink: `https://app.zapcms.com/reset-password?token=${token}`
          }
        });
      }

      return {
        message: 'A link to reset your password will be sent to your registered email.'
      };
    },
    async resetPassword(parent, { password, token }, { prisma }, info) {
      const dbUser = await prisma.user({ resetPasswordToken: token });

      if (dbUser) {
        await prisma.updateUser({
          where: { resetPasswordToken: token },
          data: {
            password: hashedPassword(password),
            resetPasswordToken: null
          }
        });

        return {
          message: 'Password updated successfully.'
        };
      }

      throw new AuthenticationError('Password reset token is invalid.');
    },
    async updateUser(parent, { firstName, lastName, email, password }, { user, prisma }, info) {
      if (password) {
        return prisma.updateUser({
          where: { id: user.id },
          data: {
            firstName,
            lastName,
            email,
            password: hashedPassword(password)
          }
        });
      }

      return prisma.updateUser({
        where: { id: user.id },
        data: {
          firstName,
          lastName,
          email
        }
      });
    },
    async deleteUser(
      parent,
      args,
      {
        user: { id },
        prisma
      },
      info
    ) {
      return prisma.deleteUser({ id });
    },
    async addCreditCard(parent, args, context, info) {
      const { stripeCustomerId: customerId } = context.user;
      const { token } = args;

      await createCard({
        token,
        customerId
      });

      return {
        message: 'Successfully updated billing information.'
      };
    },
    async subscribePlan(parent, args, context, info) {
      const { planId } = args;
      const { stripeCustomerId: customerId } = context.user;

      await createSubscription({
        planId,
        customerId
      });

      return {
        message: 'Successfully subscribed to plan.'
      };
    }
  }
};

const hashedPassword = password => bcrypt.hashSync(password, 10);
const getJwt = ({ id, email }) => jsonwebtoken.sign({ id, email }, process.env.JWT_SECRET);
