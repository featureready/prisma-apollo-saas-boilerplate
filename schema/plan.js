import { gql } from 'apollo-server-express';

export default gql`
  extend type Query {
    Plans: [Plan] @requireAuth
  }

  type Plan {
    id: String!
    title: String
    description: String
    amount: Int!
    currency: String!
    interval: String!
  }
`;
