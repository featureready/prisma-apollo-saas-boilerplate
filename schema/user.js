import { gql } from 'apollo-server-express';

export default gql`
  extend type Query {
    me: User @requireAuth
    paymentHistory: [Invoice] @requireAuth
  }

  extend type Mutation {
    signup(firstName: String!, lastName: String!, email: String!, password: String!): AuthPayload!
      @analytics(type: "identify")
    login(email: String!, password: String!): AuthPayload! @analytics(type: "identify")
    forgotPassword(email: String!): Result
    resetPassword(password: String!, token: String!): Result
    updateUser(firstName: String, lastName: String, email: String, password: String): User!
      @requireAuth
    deleteUser(id: ID!): Int!
    addCreditCard(token: String): Result @requireAuth
    subscribePlan(planId: String!): Result @requireAuth
  }

  type AuthPayload {
    jwt: String!
    user: User!
  }

  type Result {
    message: String
  }

  type User {
    id: String
    firstName: String
    lastName: String
    fullName: String @computed(value: "$firstName $lastName")
    email: String
    createdAt: DateTime
    updatedAt: DateTime
    periodStart: DateTime
    periodEnd: DateTime
    team: Team
  }

  type Invoice {
    amountDue: Int
    amountPaid: Int
    invoicePdf: String
    status: String
    date: DateTime
    periodStart: DateTime
    periodEnd: DateTime
  }
`;
