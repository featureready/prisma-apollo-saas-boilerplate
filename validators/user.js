import * as yup from 'yup';
import validate from './validate';

const schema = yup.object().shape({
  firstName: yup
    .string()
    .label('First name')
    .min(1)
    .max(15),
  lastName: yup
    .string()
    .label('Last name')
    .min(1)
    .max(15),
  email: yup
    .string()
    .email()
    .min(1),
  password: yup.string().min(6)
});

export const validateUser = input => validate({ input, schema });

export default {
  validateUser
};
