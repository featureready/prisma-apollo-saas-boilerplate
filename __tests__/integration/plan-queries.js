import { graphqlRequest, prisma } from '../../lib/test-util';
import { SIGNUP } from '../../queries/user';
import { PLANS } from '../../queries/plan';

let user;
let jwt;

beforeEach(async () => {
  ({
    data: {
      Signup: { user, jwt }
    }
  } = await graphqlRequest({
    query: SIGNUP,
    variables: {
      input: {
        firstName: 'Test',
        lastName: 'User',
        email: 'test@user.com',
        password: 'testpassword'
      }
    }
  }));
});

afterEach(async () => {
  await prisma.deleteManyUsers();
});

it('able to retrieve all plans', async () => {
  const res = await graphqlRequest({
    headers: {
      Authorization: `Bearer ${jwt}`
    },
    query: PLANS
  });

  expect(res).toMatchSnapshot();
});
