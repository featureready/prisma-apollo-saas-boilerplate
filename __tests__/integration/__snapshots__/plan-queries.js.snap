// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`able to retrieve all plans 1`] = `
Object {
  "data": Object {
    "Plans": Array [
      Object {
        "amount": 6000,
        "currency": "aud",
        "description": "10 projects, unlimited API calls",
        "id": "plan_EwhdWAyl20lY3r",
        "interval": "month",
        "title": "Agency - Monthly",
      },
      Object {
        "amount": 3000,
        "currency": "aud",
        "description": "3 projects, unlimited API calls",
        "id": "plan_EwhcP2qanL2la9",
        "interval": "month",
        "title": "Freelancer - Monthly",
      },
      Object {
        "amount": 0,
        "currency": "aud",
        "description": "1 project, 500 API calls per day",
        "id": "plan_EwhcFkDkq3Wfzr",
        "interval": "month",
        "title": "Free - Monthly",
      },
    ],
  },
}
`;
